### DISCUSSION TITLE
"Simple way" of filtering Data Buckets (Example Method)

### DISCUSSION THREAD

**Discussion by jpeter06** [mockoon/mockoon#1053] [Help]
Hello all,
Just wanted to share a method to _filter_ **Data Buckets List**. ¿Any other idea?
The problem was how to obtain the full Object in Json format.

`---route---`

```
{{setVar 'addComma' false }} 
[ 
  {{#each (dataRaw 'TEST_DATA' )}} 
    {{#if ( includes  this.type   (urlParam 'type'))}}
      {{#if @addComma}},{{/if}}
      {{data 'TEST_DATA' (concat '' @index)}}
      {{setVar 'addComma' true }}
    {{/if}}
  {{/each}}
]
```

`--- route path ---`
testRawData/:type

`---- Query with urlParam = t1 ---`
http://localhost:3000/testRawData/t1

`--- Result ---`
`[ {"id":"id0","name":"label0","type":"t1"} , {"id":"id1","name":"label1","type":"t1"} ]`

`----- Data Buckets List  TEST_DATA -----`
```
[
  {"id":"id0" ,"name": "label0","type": "t1"},
  {"id":"id1" , "name": "label1", "type": "t1"},
  {"id":"id2" ,"name": "label2", "type": "t2"}
]
```

**Comment by 255kb** [comment]
Thank you for sharing this!

**Comment by janeivind** [comment]
This is great, but I would love if this could work:

```
{{ stringify (oneOf (
    filter
      (dataRaw 'TEST_DATA')
      (object type=(urlParam 'type') )
  ))
}}
```

**Reply by 255kb** [reply]
It is working with triple curly braces (avoid Handlebars characters escaping): 
```handlebars
{{{ stringify (oneOf (
    filter
      (dataRaw 'TEST_DATA')
      (object type=(urlParam 'type') )
  ))
}}}
```
Be sure to generate a JSON valid array in the data bucket, otherwise it will be treated as a string and you won't be able to manipulate it.

**Reply by janeivind** [reply]
Great, got it working 👍 
But how about queryParam...? Trying same setup, empty result but should it work?

```
{{{ stringify (
        filter
          (dataRaw "basicElements")
          (object status=(queryParam 'status'))
      )
    }}}
```
It works if I change `(queryParam 'status')` with a hardcoded value in hyphens.

**Reply by skdutta-ga** [reply]
I have exactly the same issue

**Reply by 255kb** [reply]
You need to use `queryParamRaw`. 
I admit that the raw and non-raw helpers can be confusing. There is a bit of history here as all the helpers were initially "non raw" and returning simple strings or stringified content. It's only later I realized that to be able to use those data with helpers like `each` or `if` they would need to access the underlying data structures (JS). So I created the "raw" helpers to avoid breaking changes, but they should definitely be the default. 

And while `urlParam` is always containing a string, [query parameters can be interpreted as arrays and objects](https://mockoon.com/docs/latest/api-endpoints/routing/#query-parameters) (with dot or bracket notation). That's why using the "raw" version is required.

**Reply by HighLiuk** [reply]
I'd love to use that on CRUD routes as a preliminary filtering condition. Otherwise I would need to reimplement pagination and sorting if I want filtering.
And if I want to reinvent the wheel, it's not actually possible to sort array results based other than using CRUD routes right?
The best thing would be if CRUD routes supported filters.
Looks like it is not much complicated by editing the CRUD function in the source code, is it?

**Reply by 255kb** [reply]
@HighLiuk Yes, it would be better as there is no easy way to sort, like you said. 
It's not a very complicated task indeed, but it still raises some questions (do we rename the current query parameters sort, limit, order with an underscore, to avoid conflicts, do we keep it like this, what about query params containing arrays, etc)
We can follow up on this [opened issue](https://github.com/mockoon/mockoon/issues/1047).

**Comment by janeivind** [comment]
Some examples on data bucket with [JSONPath Plus](https://jsonpath-plus.github.io/JSONPath/docs/ts/):

Query Param name equals:
`
{{ data 'TEST_DATA' (concat '$.[?(@.name =="' (queryParamRaw "name") '")]')  }}
`

Query Param name matches:
`
{{ data 'TEST_DATA' (concat '$.[?(@property === "name" && @.match(/' (queryParamRaw "name") '/i))]^') }}
`

**Comment by HighLiuk** [comment]
Here there's a snippet for full text search that works for data that have objects with primitive values:

```
{{setVar 'query' (concat '$[?(/' (queryParam 'q') '/i.test(Object.values(@).join(" ")))]')}}
{{dataRaw 'TEST_DATA' @query}}
```
